pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go

```diff
@@ -0,0 +1,282 @@
+			c.bigThreadJunks[0] = remaining
+		} else {
+			c.bigThreadJunks = c.bigThreadJunks[1:]
+			c.smallJunks = append(c.smallJunks, remaining...)
+		}
+		return &thread
+	}
+	return nil
+}
+
+func (c *cell) IsEmpty() bool {
+	return len(c.smallJunks) == 0 && len(c.bigThreadJunks) == 0
+}
+
+type cpuPool struct {
+	cells             []*cell
+	threadsPerCore    int
+	cores             int
+	allowCellCrossing bool
+}
+
+func NewStrictCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int) VCPUPool {
+	return newCPUPool(requestedToplogy, nodeTopology, cpuSet, false)
+}
+
+func NewRelaxedCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int) VCPUPool {
+	return newCPUPool(requestedToplogy, nodeTopology, cpuSet, true)
+}
+
+func newCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int, allowCellCrossing bool) *cpuPool {
+	pool := &cpuPool{threadsPerCore: int(requestedToplogy.Threads), cores: int(requestedToplogy.Cores * requestedToplogy.Sockets), allowCellCrossing: allowCellCrossing}
+	cores := cpuJunksToCells(cpuSet, nodeTopology)
+
+	for _, coresOnCell := range cores {
+		c := cell{threadsPerCore: int(requestedToplogy.Threads)}
+		for j, core := range coresOnCell {
+			if len(core) >= c.threadsPerCore {
+				c.bigThreadJunks = append(c.bigThreadJunks, coresOnCell[j])
+			} else {
+				c.smallJunks = append(c.smallJunks, coresOnCell[j]...)
+			}
+		}
+		pool.cells = append(pool.cells, &c)
+	}
+	return pool
+}
+
+// cpuJunksToCells takes the allocated cpuset, determines which of the threads belongs to which cpu and which numa
+// cell and returns an aggregated view. The return map has the numa caell as index and contains for each cpu an array
+// of all allocated threads.
+func cpuJunksToCells(cpuSet []int, nodeTopology *v1.Topology) (cores [][][]uint32) {
+	threads := map[int]struct{}{}
+	visited := map[uint32]struct{}{}
+	cores = [][][]uint32{}
+	for _, cpu := range cpuSet {
+		threads[cpu] = struct{}{}
+	}
+	for _, cell := range nodeTopology.NumaCells {
+		var coresOnCell [][]uint32
+		for _, cpu := range cell.Cpus {
+			if _, exists := visited[cpu.Id]; exists {
+				continue
+			}
+			core := []uint32{}
+			if len(cpu.Siblings) == 0 {
+				visited[cpu.Id] = struct{}{}
+				if _, exists := threads[int(cpu.Id)]; exists {
+					core = append(core, cpu.Id)
+				}
+			} else {
+				for _, thread := range cpu.Siblings {
+					visited[thread] = struct{}{}
+					if _, exists := threads[int(thread)]; exists {
+						core = append(core, thread)
+					}
+				}
+			}
+			coresOnCell = append(coresOnCell, core)
+		}
+		cores = append(cores, coresOnCell)
+	}
+	return cores
+}
+
+func (p *cpuPool) FitCores() (cpuTune *api.CPUTune, err error) {
+	threads, remaining := p.fitCores(p.cores)
+
+	if remaining > 0 {
+		if p.allowCellCrossing {
+			return nil, fmt.Errorf("not enough exclusive threads provided, could not fit %v core(s)", remaining)
```

Comment:
what would be a scenario where allowCellCrossing == true but remaining cores > 0.  Is this only possible when someone requests threads per core where the number of threads can't possibly match the number of cores?

I'm just trying to understand if this is something that can be caught in validation webhook.

Reply:
This is one of the main issues with all this. We don't know what we get until the kubelet is done and everything is started. We do not know up to this point if we can map cpu threads in a reasonable way. Note that for cpu pinning (so no numa mapping), we should always get a working pinning but it can be inefficient. For numa passthrough it is more likely that it will fail, because there are cases where we can't create a correct mapping out of the assigned CPUs because we can't form via libvirt topologies where cpus have different amounts of threads.